### PR TITLE
Add OpenAPI V3 validation to app CRD

### PR DESCRIPTION
**Pull request by rossf7**
Towards https://github.com/giantswarm/giantswarm/issues/5407

Adds validation for the app CRD and aims to align with the release group CRDs.

### CODE REVIEW

File: pkg/apis/application/v1alpha1/app_types.go [hunk 1]
```diff
@@ -9,21 +10,214 @@ const (
...
+	OpenAPIV3Schema: &apiextensionsv1beta1.JSONSchemaProps{
+		Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+			"spec": {
+				Type: "object",
+				Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
+					"catalog": {
+						Type:      "string",
+						MinLength: to.Int64P(3),
+					},
+					"config": {
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: I have a doubt about optional properties. The config and secret blocks are both optional. Do I need to define that somehow?

File: pkg/apis/application/v1alpha1/app_types.go [hunk 2]
```diff
@@ -9,21 +10,214 @@ const (
...
+//   group: application.giantswarm.io
+//   names:
+//     kind: App
+//     listKind: AppList
+//     plural: apps
+//     singular: app
+//   scope: Namespaced
+//   subresources:
+//     status: {}
+//   validation:
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_types.go]: The example CRD is pretty huge now. Should we include the validation?

File: pkg/apis/application/v1alpha1/app_types.go [hunk 3]
```diff
@@ -9,21 +10,214 @@ const (
...
+//                     name:
+//                       minLength: 3
+//                       type: string
+//                     namespace:
+//                       minLength: 3
+//                       type: string
+//                   type: object
+//               type: object
+//             version:
+//               pattern: ^\d+\.\d+\.\d+$
```
Comment (kopiczko) [inline pkg/apis/application/v1alpha1/app_types.go]: I think this is too strict. What about `1.2.3-mywodnerfulgitsha12345`?

**Review by kopiczko — COMMENTED** [review]
I think in general mid term it would be good to parse text in runtime instead of generating text from structs. Those structs for validation are nightmare.

File: pkg/apis/application/v1alpha1/app_types.go (hunk 3 above)
Comment (rossf7) [reply pkg/apis/application/v1alpha1/app_types.go]: Good catch. I'll make it more permissive.

**Review by rossf7 — COMMENTED** [review]
> Those structs for validation are nightmare.

Agreed. So the master copy would be in YAML? Where would that be stored? Could we use a constant like we do for the current struct?

**Comment by kopiczko** [comment]
> > Those structs for validation are nightmare.
> 
> Agreed. So the master copy would be in YAML? Where would that be stored? Could we use a constant like we do for the current struct?

I think go string const is by far more readable. We have that in the comment already which we can replace with the const. If the const is parsed in `init` functions we'll catch errors/panics in the unit test phase of the CI. WDYT?

**Comment by rossf7** [comment]
OK cool thanks. I'll update this to use a string const. Yes I think using an init function works.

We started working on this during SWH but it's also part of our team story.

cc @sslavic

**Review by kopiczko — APPROVED** [review]

**Comment by kopiczko** [comment]
LGTM when version validation is relaxed.

File: pkg/apis/application/v1alpha1/app_types.go (hunk 1 above)
Comment (sslavic) [reply pkg/apis/application/v1alpha1/app_types.go]: From https://swagger.io/docs/specification/data-models/data-types/#required

> By default, all object properties are optional. You can specify the required properties in the required list:

Comment (rossf7) [reply pkg/apis/application/v1alpha1/app_types.go]: @sslavic Ack and thanks that link is super useful.

**Comment by rossf7** [comment]
Closing in favor of #279.

**Pull request closed without merging**